README.md
Pilbox [![Build Status](https://travis-ci.org/agschwender/pilbox.png)](https://travis-ci.org/agschwender/pilbox) [![Coverage Status](https://coveralls.io/repos/agschwender/pilbox/badge.png)](https://coveralls.io/r/agschwender/pilbox)
======

Pilbox is an image resizing application server built on Python's [Tornado web framework](http://www.tornadoweb.org/en/stable/) using the [Python Imaging Library (Pillow)](https://pypi.python.org/pypi/Pillow/2.1.0). It is not intended to be the primary source of images, but instead acts as a proxy which requests images and resizes them as desired.

Setup
=====

Dependencies
------------

  * >= [Python 2.7](http://www.python.org/download/)
  * [Pillow 2.1.0](https://pypi.python.org/pypi/Pillow/2.1.0)
  * [Tornado 3.1](https://pypi.python.org/pypi/tornado/3.1)
  * [OpenCV 2.x](http://opencv.org/) (optional)
  * Image Libraries: libjpeg-dev, libfreetype6-dev, libwebp-dev, zlib1g-dev

Vagrant
-------

Packaged with Pilbox is a [Vagrant](http://www.vagrantup.com/) configuration file which installs all necessary dependencies on a virtual box using [Ansible](http://www.ansibleworks.com/). See the [Vagrant documentation](http://docs.vagrantup.com/v2/installation/) and the [Ansible documentation](http://www.ansibleworks.com/docs/gettingstarted.html#getting-ansible) for installation instructions. Once installed, the following will start and provision a virtual machine.

    $ vagrant up
    $ vagrant provision

To access the virtual machine itself, simply...

    $ vagrant ssh

Running
=======

Manual
------

To run the application, issue the following command

    $ python -m pilbox.app

By default, this will run the application on port 8888 and can be accessed by visiting:

    http://localhost:8888/

To see a list of all available options, run

    $ python -m pilbox.app --help
    Usage: pilbox/app.py [OPTIONS]

    Options:

      --allowed_hosts            list of allowed hosts (default [])
      --background               default hexadecimal bg color (RGB or ARGB)
      --client_key               client key
      --client_name              client name
      --config                   path to configuration file
      --debug                    run in debug mode (default False)
      --expand                   default to expand when rotating
      --filter                   default filter to use when resizing
      --help                     show this help information
      --max_requests             max concurrent requests (default 40)
      --port                     run on the given port (default 8888)
      --position                 default cropping position
      --quality                  default jpeg quality, 0-100
      --timeout                  timeout of requests in seconds (default 10)

Vagrant
-------

When running via Vagrant, the application is automatically started on port 8888 on 192.168.100.100, i.e.

    http://192.168.100.100:8888/

Calling
=======

To use the image resizing service, include the application url as you would any other image. E.g. this image url

```html
<img src="http://i.imgur.com/zZ8XmBA.jpg" />
```

Would be replaced with this image url

```html
<img src="http://localhost:8888/?url=http%3A%2F%2Fi.imgur.com%2FzZ8XmBA.jpg&w=300&h=300&mode=crop" />
```

This will request the image served at the supplied url and resize it to `300x300` using the `crop` mode. The below is the list of parameters that can be supplied to the service.

### General Parameters

  * _url_: The url of the image to be resized
  * _op_: The operation to perform: noop, resize (default), rotate
    * _noop_: No operation is performed, image is returned as it is received
    * _resize_: Resize the image
    * _rotate_: Rotate the image

### Resize Parameters

  * _w_: The desired width of the image
  * _h_: The desired height of the image
  * _mode_: The resizing method: clip, crop (default), fill and scale
    * _clip_: Resize to fit within the desired region, keeping aspect ratio
    * _crop_: Resize so one dimension fits within region, center, cut remaining
    * _fill_: Fills the clipped space with a background color
    * _scale_: Resize to fit within the desired region, ignoring aspect ratio
  * _bg_: Background color used with fill mode (RGB or ARGB)
    * _RGB_: 3- or 6-digit hexadecimal number
    * _ARGB_: 4- or 8-digit hexadecimal number, only relevant for PNG images
  * _filter_: The filtering algorithm used for resizing
    * _nearest_: Fastest, but often images appear pixelated
    * _bilinear_: Faster, can produce acceptable results
    * _bicubic_: Fast, can produce acceptable results
    * _antialias_: Slower, produces the best results
  * _fmt_: The output format to save as, defaults to the source format
    * _jpeg_: Save as JPEG
    * _png_: Save as PNG
    * _webp_: Save as WebP
  * _pos_: The crop position
    * _top-left_: Crop from the top left
    * _top_: Crop from the top center
    * _top-right_: Crop from the top right
    * _left_: Crop from the center left
    * _center_: Crop from the center
    * _right_: Crop from the center right
    * _bottom-left_: Crop from the bottom left
    * _bottom_: Crop from the bottom center
    * _bottom-right_: Crop from the bottom right
    * _face_: Identify faces and crop from the midpoint of their position(s)
    * _x,y_: Custom center point position ratio, e.g. 0.0,0.75
  * _q_: The quality (1-100) used to save the image, only relevant to JPEGs.

### Rotate Parameters

  * _deg_: The desired rotation angle degrees
  * _expand_: Expand the sizeto include the full rotated image
  * _fmt_: The output format to save as, defaults to the source format
    * _jpeg_: Save as JPEG
    * _png_: Save as PNG
    * _webp_: Save as WebP
  * _q_: The quality (1-100) used to save the image, only relevant to JPEGs.

### Security-related Parameters

  * _client_: The client name
  * _sig_: The signature

The `url` parameter is always required as it dictates the image that will be manipulated. `op` is optional and defaults to `resize`. It also supports a comma separated list of operations, where each operation is applied in the order that it appears in the list. Depending on the operation, additional parameters are required.  All image manipulation requests accept `fmt` and `q`. `fmt` is optional and defaults to the source image format. `q` is optional and defaults to `90`. To ensure security, all requests also support, `client` and `sig`. `client` is required only if the `client_name` is defined within the configuration file. Likewise, `sig` is required only if the `client_key` is defined within the configuration file. See the [signing section](#signing) for details on how to generate the signature.

For resizing, either the `w` or `h` parameter is required. If only one dimension is specified, the application will determine the other dimension using the aspect ratio. `mode` is optional and defaults to `crop`. `filter` is optional and defaults to `antialias`.  `bg` is optional and defaults to `fff`. `pos` is optional and defaults to `center`.

For rotating, `deg` is required. `expand` is optional and defaults to `0` (disabled). It is recommended that this feature not be used as it typically does not produce high quality images.

Note, all built-in defaults can be overridden by setting them in the configuration file. See the [configuration section](#configuration) for more details.

Examples
========

The following images show the various resizing modes in action for an original image size of `640x428` that is being resized to `500x400`.

Clip
----

The image is resized to fit within a `500x400` box, maintaining aspect ratio and producing an image that is `500x334`. Clipping is useful when no portion of the image can be lost and it is acceptable that the image not be exactly the supplied dimensions, but merely fit within the dimensions.

![Clipped image](pilbox/test/data/expected/example-500x400-clip.jpg)

Crop
----

The image is resized so that one dimension fits within the `500x400` box. It is then centered and the excess is cut from the image. Cropping is useful when the position of the subject is known and the image must be exactly the supplied size.

![Cropped image](pilbox/test/data/expected/example-500x400-crop.jpg)

Fill
----

Similar to clip, fill resizes the image to fit within a `500x400` box. Once clipped, the image is centered within the box and all left over space is filled with the supplied background color. Filling is useful when no portion of the image can be lost and it must be exactly the supplied size.

![Filled image](pilbox/test/data/expected/example-500x400-fill-ccc.jpg)

Scale
-----

The image is clipped to fit within the `500x400` box and then stretched to fill the excess space. Scaling is often not useful in production environments as it generally produces poor quality images. This mode is largely included for completeness.

![Scale image](pilbox/test/data/expected/example-500x400-scale.jpg)


Testing
=======

To run all tests, issue the following command

    $ python -m pilbox.test.runtests

To run individual tests, simply indicate the test to be run, e.g.

    $ python -m pilbox.test.runtests pilbox.test.signature_test

Signing
=======

In order to secure requests so that unknown third parties cannot easily use the resize service, the application can require that requests provide a signature. To enable this feature, set the `client_key` option. The signature is a hexadecimal digest generated from the client key and the query string using the HMAC-SHA1 message authentication code (MAC) algorithm. The below python code provides an example implementation.

```python
import hashlib
import hmac

def derive_signature(key, qs):
    m = hmac.new(key, None, hashlib.sha1)
    m.update(qs)
    return m.hexdigest()
```

The signature is passed to the application by appending the `sig` parameter to the query string; e.g. `x=1&y=2&z=3&sig=c9516346abf62876b6345817dba2f9a0c797ef26`. Note, the application does not include the leading question mark when verifying the supplied signature. To verify your signature implementation, see the `pilbox.signature` command described in the [tools section](#tools).

Configuration
=============

All options that can be supplied to the application via the command line, can also be specified in the configuration file. Configuration files are simply python files that define the options as variables. The below is an example configuration.

    # General settings
    port = 8888

    # Set client name and key if the application requires signed requests. The
    # client must sign the request using the client_key, see README for
    # instructions.
    client_name = "sample"
    client_key = "3NdajqH8mBLokepU4I2Bh6KK84GUf1lzjnuTdskY"

    # Set the allowed hosts as an alternative to signed requests. Only those
    # images which are served from the following hosts will be requested.
    allowed_hosts = ["localhost"]

    # Request-related settings
    max_requests = 50
    timeout = 7.5

    # Set default resizing options
    background = "ccc"
    filter = "bilinear"
    mode = "crop"
    position = "top"

    # Set default rotating options
    expand = False

    # Set default saving options
    format = None
    quality = 90

Tools
=====

To verify that your client application is generating correct signatures, use the signature command.

    $ python -m pilbox.signature --key=abcdef "x=1&y=2&z=3"
    Query String: x=1&y=2&z=3
    Signature: c9516346abf62876b6345817dba2f9a0c797ef26
    Signed Query String: x=1&y=2&z=3&sig=c9516346abf62876b6345817dba2f9a0c797ef26

The application allows the use of the resize functionality via the command line.

    $ python -m pilbox.image --width=300 --height=300 http://i.imgur.com/zZ8XmBA.jpg > /tmp/foo.jpg

If a new mode is added or a modification was made to the libraries that would change the current expected output for tests, run the generate test command to regenerate the expected output for the test cases.

    $ python -m pilbox.test.genexpected

Deploying
=========

The application itself does not include any caching. It is recommended that the application run behind a CDN for larger applications or behind varnish for smaller ones.

Defaults for the application have been optimized for quality rather than performance. If you wish to get higher performance out of the application, it is recommended you use a less computationally expensive filtering algorithm and a lower JPEG quality. For example, add the following to the configuration.

    # Set default resizing options
    filter = "bicubic"
    quality = 75

Changelog
=========

  * 0.1: Image resizing fit
  * 0.1.1: Image cropping
  * 0.1.2: Image scaling
  * 0.2: Configuration integration
  * 0.3: Signature generation
  * 0.3.1: Signature command-line tool
  * 0.4: Image resize command-line tool
  * 0.5: Facial recognition cropping
  * 0.6: Fill resizing mode
  * 0.7: Resize using crop position
  * 0.7.1: Resize using a single dimension, maintaining aspect ratio
  * 0.7.2: Added filter and quality options
  * 0.7.3: Support python 3
  * 0.7.4: Fixed cli for image generation
  * 0.7.5: Write output in 16K blocks
  * 0.8: Added support for ARGB (alpha-channel)
  * 0.8.1: Increased max clients and write block sizes
  * 0.8.2: Added configuration for max clients and timeout
  * 0.8.3: Only allow http and https protocols
  * 0.8.4: Added support for WebP
  * 0.8.5: Added format option and configuration overrides for mode and format
  * 0.8.6: Added custom position support
  * 0.9: Added rotate operation

Docker
======

Experimental support for [Docker](http://www.docker.io/). To build a docker container and provision it, issue the following:

    $ docker build .
    $ docker run -d -p 2222:22 -v `pwd`:/pilbox -t <imageid> ssh
    $ ansible-playbook -i provisioning/docker provisioning/playbook.yml
    $ docker stop $(docker ps -a -q)
    $ docker run -i -p :80 -p :8080 -p :8888 -v `pwd`:/pilbox -t <imageid> web

SaltStack
=========

SaltStack state for pilbox.
Replace Ansible provisioning with salt privisioning in Vagrantfile::

    config.vm.synced_folder "provisioning/salt/roots/", "/srv/salt"
    config.vm.synced_folder "provisioning/files/", "/srv/salt/files"
    config.vm.provision :salt do |salt|
      salt.minion_config = "provisioning/salt/minion"
      salt.run_highstate = true
    end

TODO
====

  * How to reconcile unavailable color profiles?
  * Add backends (S3, file system, etc...) if necessary


CHANGES.txt
  * 0.1: Image resizing fit
  * 0.1.1: Image cropping
  * 0.1.2: Image scaling
  * 0.2: Configuration integration
  * 0.3: Signature generation
  * 0.3.1: Signature command-line tool
  * 0.4: Image resize command-line tool
  * 0.5: Facial recognition cropping
  * 0.6: Fill resizing mode
  * 0.7: Resize using crop position
  * 0.7.1: Resize using a single dimension, maintaining aspect ratio
  * 0.7.2: Added filter and quality options
  * 0.7.3: Support python 3
  * 0.7.4: Fixed cli for image generation
  * 0.7.5: Write output in 16K blocks
  * 0.8: Added support for ARGB (alpha-channel)
  * 0.8.1: Increased max clients and write block sizes
  * 0.8.2: Added configuration for max clients and timeout
  * 0.8.3: Only allow http and https protocols
  * 0.8.4: Added support for WebP
  * 0.8.5: Added format option and configuration overrides for mode and format
  * 0.8.6: Added custom position support
  * 0.9: Added rotate operation
  * 0.9.1: Added sub-region selection operation
  * 0.9.4: Added Pilbox as a PyPI package
  * 0.9.10: Converted README to reStructuredText
  * 0.9.14: Added Sphinx docs
  * 0.9.15: Added implicit base url
  * 0.9.16: Added validate cert to configuration
  * 0.9.17: Added support for GIF format
  * 0.9.18: Fix for travis builds on python 2.6 and 3.3
  * 0.9.19: Validate cert fix
  * 0.9.20: Added optimize option
  * 0.9.21: Added console script entry point
  * 1.0.0: Modified for easier library usage
  * 1.0.1: Added allowed operations and default operation
  * 1.0.2: Modified to allow override of http content type
  * 1.0.3: Safely catch image save errors
  * 1.0.4: Added progressive option
  * 1.1.0: Proxy server support
  * 1.1.1: Added JPEG auto rotation based on Exif orientation
  * 1.1.2: Added keep JPEG quality option and set JPEG subsampling to keep
  * 1.1.3: Fix auto rotation on JPEG with missing Exif data
  * 1.1.4: Exception handling around invalid Exif data
  * 1.1.5: Fixed image requests without content types
  * 1.1.6: Support custom applications that need command line arguments
  * 1.1.7: Support adapt resize mode
  * 1.1.8: Added preserve Exif flag
  * 1.1.9: Increased Pillow version to 2.8.1
  * 1.1.10: Added ca_certs option
  * 1.1.11: Added support for TIFF
  * 1.2.0: Support setting background when saving a transparent image
    - *Backwards incompatible*: default background property changed to `0fff`. To restore previous behavior, set background in config to `ffff`.
  * 1.2.1: Added max operations config property
  * 1.2.2: Added max resize width and height config properties
  * 1.2.3: Added user_agent option
  * 1.3.0: Increased Pillow to 4.1.0 and Tornado to 4.5.1
  * 1.3.1: Fix pilbox.image CLI for python 3.0
  * 1.3.2: Fix GIF P-mode to JPEG conversion
  * 1.3.3: Increase Pillow version to 5.2.0 and Tornado version to 5.1.0
  * 1.3.4: Added worker config property to set number of Tornado processes


pilbox/test/gif_test.py
from __future__ import absolute_import, division, with_statement

import itertools
import os
import os.path
import re
from sys import platform as _platform
from io import BytesIO

from PIL import Image, ImageChops, ImageSequence
from PIL.GifImagePlugin import getheader, getdata

import pilbox.image
from tornado.test.util import unittest

DATADIR = os.path.join(os.path.dirname(__file__), "data")
ANIMATED_GIF = os.path.join(DATADIR, "animated.gif")
EXPECTED_GIF = os.path.join(DATADIR, "expected", "animated.gif")

# stolen from https://github.com/python-pillow/Pillow/blob/master/Scripts/gifmaker.py
def write_gif(fp, sequence):
    previous = None
    for im in sequence:
        duration = im.info['duration']
        loop = im.info['loop']
        im = resize(im)
        if not previous:
            for s in getheader(im)[0] + getdata(im, duration=duration, loop=loop):
                fp.write(s)
        else:
            delta = ImageChops.subtract_modulo(im, previous)
            bbox = delta.getbbox()
            if bbox:
                for s in getdata(im.crop(bbox), offset=bbox[:2], duration=duration, loop=loop):
                    fp.write(s)
            else:
                pass
        previous = im.copy()
    fp.write(";".encode('utf-8'))


def resize(im, width=200, height=100):
    outfile = BytesIO()
    im.save(outfile, 'gif')
    image = pilbox.image.Image(outfile)
    image.resize(width, height)
    return image.img


class GifTest(unittest.TestCase):

    def test_resize(self):
        with open(EXPECTED_GIF, "wb") as fp:
            image = Image.open(ANIMATED_GIF)
            frames = ImageSequence.Iterator(image)
            write_gif(fp, frames)


pilbox/test/runtests.py
#!/usr/bin/env python

from __future__ import absolute_import, division, with_statement

import logging
import sys
import textwrap

from tornado.test.util import unittest

TEST_MODULES = [
    'pilbox.test.app_test',
    'pilbox.test.errors_test',
    'pilbox.test.image_test',
    'pilbox.test.gif_test',
    'pilbox.test.signature_test',
]


def all():
    return unittest.defaultTestLoader.loadTestsFromNames(TEST_MODULES)


class PilboxTextTestRunner(unittest.TextTestRunner):
    def run(self, test):
        result = super(PilboxTextTestRunner, self).run(test)
        if result.skipped:
            skip_reasons = set(reason for (test, reason) in result.skipped)
            self.stream.write(textwrap.fill(
                "Some tests were skipped because: %s" %
                ", ".join(sorted(skip_reasons))))
            self.stream.write("\n")
        return result


if __name__ == '__main__':
    import warnings
    # Be strict about most warnings.  This also turns on warnings that are
    # ignored by default, including DeprecationWarnings and
    # python 3.2's ResourceWarnings.
    warnings.filterwarnings("error")
    warnings.filterwarnings("ignore", category=ImportWarning)
    warnings.filterwarnings("ignore", category=DeprecationWarning)
    warnings.filterwarnings("error", category=DeprecationWarning,
                            module=r"tornado\..*")
    # The unittest module is aggressive about deprecating redundant methods,
    # leaving some without non-deprecated spellings that work on both
    # 2.7 and 3.2
    warnings.filterwarnings("ignore", category=DeprecationWarning,
                            message="Please use assert.* instead")

    logging.getLogger("tornado.access").setLevel(logging.CRITICAL)

    import tornado.testing
    kwargs = {}
    if sys.version_info >= (3, 2):
        # HACK:  unittest.main will make its own changes to the warning
        # configuration, which may conflict with the settings above
        # or command-line flags like -bb.  Passing warnings=False
        # suppresses this behavior, although this looks like an implementation
        # detail.  http://bugs.python.org/issue15626
        kwargs['warnings'] = False
    kwargs['testRunner'] = PilboxTextTestRunner
    tornado.testing.main(**kwargs)


pilbox/app.py
#!/usr/bin/env python
#
# Copyright 2013 Adam Gschwender
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

from __future__ import absolute_import, division, with_statement

import logging
import socket

import tornado.escape
import tornado.gen
import tornado.httpclient
import tornado.httpserver
import tornado.ioloop
import tornado.options
import tornado.web
from tornado.options import define, options, parse_config_file

from pilbox import errors
from pilbox.image import Image
from pilbox.signature import verify_signature

try:
    from urlparse import urlparse, urljoin
except ImportError:
    from urllib.parse import urlparse, urljoin

try:
    import pycurl
except ImportError:
    pycurl = None


# general settings
define("config", help="path to configuration file",
       callback=lambda path: parse_config_file(path, final=False))
define("debug", help="run in debug mode", type=bool, default=False)
define("port", help="run on the given port", type=int, default=8888)
define("workers", help="number of worker processes (0 = auto)",
       type=int, default=0)

# security related settings
define("client_name", help="client name")
define("client_key", help="client key")
define("allowed_hosts", help="valid hosts", default=[], multiple=True)
define("allowed_operations", help="valid ops", default=[], multiple=True)
define("max_operations", help="maximum operations to perform", default=10)
define("max_resize_height", help="maximum resize height", default=15000)
define("max_resize_width", help="maximum resize width", default=15000)

# request related settings
define("max_requests", help="max concurrent requests", type=int, default=40)
define("timeout", help="request timeout in seconds", type=float, default=10)
define("implicit_base_url", help="prepend protocol/host to url paths")
define("ca_certs",
       help="override filename of CA certificates in PEM format",
       default=None)
define("validate_cert", help="validate certificates", type=bool, default=True)
define("proxy_host", help="proxy hostname")
define("proxy_port", help="proxy port", type=int)
define("user_agent", help="user agent", type=str)

# header related settings
define("content_type_from_image",
       help="override content type using image mime type",
       type=bool)

# default image option settings
define("background", help="default hexadecimal bg color (RGB or ARGB)")
define("expand", help="default to expand when rotating", type=int)
define("filter", help="default filter to use when resizing")
define("format", help="default format to use when outputting")
define("mode", help="default mode to use when resizing")
define("operation", help="default operation to perform")
define("optimize", help="default to optimize when saving", type=int)
define("position", help="default cropping position")
define("progressive", help="default to progressive when saving", type=int)
define("quality", help="default jpeg quality, 1-99 or keep")
define("retain", help="default adaptive retain percent, 1-99", type=int)
define("preserve_exif", help="default behavior for exif data", type=int)

logger = logging.getLogger("tornado.application")


class PilboxApplication(tornado.web.Application):

    def __init__(self, **kwargs):
        settings = dict(
            debug=options.debug,
            client_name=options.client_name,
            client_key=options.client_key,
            allowed_hosts=options.allowed_hosts,
            allowed_operations=set(
                options.allowed_operations or ImageHandler.OPERATIONS),
            max_operations=options.max_operations,
            max_resize_height=options.max_resize_height,
            max_resize_width=options.max_resize_width,
            background=options.background,
            expand=options.expand,
            filter=options.filter,
            format=options.format,
            mode=options.mode,
            operation=options.operation,
            optimize=options.optimize,
            position=options.position,
            progressive=options.progressive,
            quality=options.quality,
            max_requests=options.max_requests,
            timeout=options.timeout,
            implicit_base_url=options.implicit_base_url,
            ca_certs=options.ca_certs,
            user_agent=options.user_agent,
            validate_cert=options.validate_cert,
            content_type_from_image=options.content_type_from_image,
            proxy_host=options.proxy_host,
            proxy_port=options.proxy_port,
            preserve_exif=options.preserve_exif)

        settings.update(kwargs)

        if settings.get("proxy_host") and pycurl is None:  # pragma: no cover
            raise Exception("PycURL is required for proxy requests")

        if pycurl is not None:  # pragma: no cover
            tornado.httpclient.AsyncHTTPClient.configure(
                "tornado.curl_httpclient.CurlAsyncHTTPClient")

        tornado.web.Application.__init__(self, self.get_handlers(), **settings)

    def get_handlers(self):
        return [(r"/", ImageHandler)]


class ImageHandler(tornado.web.RequestHandler):
    FORWARD_HEADERS = ["Cache-Control", "Expires", "Last-Modified"]
    OPERATIONS = ["region", "resize", "rotate", "noop"]

    _FORMAT_TO_MIME = {
        "gif": "image/gif",
        "jpeg": "image/jpeg",
        "jpg": "image/jpeg",
        "png": "image/png",
        "webp": "image/webp",
        "tiff": "image/tiff",
    }

    @tornado.gen.coroutine
    def get(self):
        self.validate_request()
        resp = yield self.fetch_image()
        self.render_image(resp)

    def get_argument(self, name, default=None, strip=True):
        return super(ImageHandler, self).get_argument(name, default, strip)

    def validate_request(self):
        self._validate_operation()
        self._validate_url()
        self._validate_signature()
        self._validate_client()
        self._validate_host()

        opts = self._get_save_options()
        ops = self._get_operations()
        if "resize" in ops:
            w, h = self.get_argument("w"), self.get_argument("h")
            Image.validate_dimensions(w, h)
            if w and int(w) > self.settings.get("max_resize_width"):
                raise errors.DimensionsError("Exceeds maximum allowed width")
            elif h and int(h) > self.settings.get("max_resize_height"):
                raise errors.DimensionsError("Exceeds maximum allowed height")
            opts.update(self._get_resize_options())
        if "rotate" in ops:
            Image.validate_degree(self.get_argument("deg"))
            opts.update(self._get_rotate_options())
        if "region" in ops:
            Image.validate_rectangle(self.get_argument("rect"))

        Image.validate_options(opts)

    @tornado.gen.coroutine
    def fetch_image(self):
        url = self.get_argument("url")
        if self.settings.get("implicit_base_url") \
                and urlparse(url).hostname is None:
            url = urljoin(self.settings.get("implicit_base_url"), url)

        client = tornado.httpclient.AsyncHTTPClient(
            max_clients=self.settings.get("max_requests"))
        try:
            resp = yield client.fetch(
                url,
                request_timeout=self.settings.get("timeout"),
                ca_certs=self.settings.get("ca_certs"),
                validate_cert=self.settings.get("validate_cert"),
                user_agent=self.settings.get("user_agent"),
                proxy_host=self.settings.get("proxy_host"),
                proxy_port=self.settings.get("proxy_port"))
            raise tornado.gen.Return(resp)
        except (socket.gaierror, tornado.httpclient.HTTPError) as e:
            logger.warn("Fetch error for %s: %s",
                        self.get_argument("url"),
                        str(e))
            raise errors.FetchError()

    def render_image(self, resp):
        outfile, outfile_format = self._process_response(resp)
        self._set_headers(resp.headers, outfile_format)
        for block in iter(lambda: outfile.read(65536), b""):
            self.write(block)
        outfile.close()

    def write_error(self, status_code, **kwargs):
        err = kwargs["exc_info"][1] if "exc_info" in kwargs else None
        if isinstance(err, errors.PilboxError):
            self.set_header("Content-Type", "application/json")
            resp = dict(status_code=status_code,
                        error_code=err.get_code(),
                        error=err.log_message)
            self.finish(tornado.escape.json_encode(resp))
        else:
            super(ImageHandler, self).write_error(status_code, **kwargs)

    def _process_response(self, resp):
        ops = self._get_operations()
        if "noop" in ops:
            return (resp.buffer, None)

        image = Image(resp.buffer)
        for operation in ops:
            if operation == "resize":
                self._image_resize(image)
            elif operation == "rotate":
                self._image_rotate(image)
            elif operation == "region":
                self._image_region(image)

        return (self._image_save(image), image.img.format)

    def _image_region(self, image):
        image.region(self.get_argument("rect").split(","))

    def _image_resize(self, image):
        opts = self._get_resize_options()
        image.resize(self.get_argument("w"), self.get_argument("h"), **opts)

    def _image_rotate(self, image):
        opts = self._get_rotate_options()
        image.rotate(self.get_argument("deg"), **opts)

    def _image_save(self, image):
        opts = self._get_save_options()
        return image.save(**opts)

    def _set_headers(self, headers, file_format):
        if file_format and any((self.get_argument("fmt"),
                                self.settings.get("format"),
                                self.settings.get("content_type_from_image"))):
            self.set_header(
                "Content-Type", self._FORMAT_TO_MIME.get(file_format.lower()))
        elif "Content-Type" in headers:
            self.set_header("Content-Type", headers["Content-Type"])

        for k in ImageHandler.FORWARD_HEADERS:
            if k in headers and headers[k]:
                self.set_header(k, headers[k])

    def _get_operations(self):
        return self.get_argument(
            "op", self.settings.get("operation") or "resize").split(",")

    def _get_resize_options(self):
        return self._get_options(
            dict(mode=self.get_argument("mode"),
                 filter=self.get_argument("filter"),
                 position=self.get_argument("pos"),
                 background=self.get_argument("bg"),
                 retain=self.get_argument("retain")))

    def _get_rotate_options(self):
        return self._get_options(
            dict(expand=self.get_argument("expand")))

    def _get_save_options(self):
        return self._get_options(
            dict(format=self.get_argument("fmt"),
                 optimize=self.get_argument("opt"),
                 quality=self.get_argument("q"),
                 progressive=self.get_argument("prog"),
                 background=self.get_argument("bg"),
                 preserve_exif=self.get_argument("exif")))

    def _get_options(self, opts):
        for k, v in opts.items():
            if v is None:
                opts[k] = self.settings.get(k, None)
        return opts

    def _validate_operation(self):
        operations = set(self._get_operations())
        if not operations.issubset(self.settings.get("allowed_operations")):
            raise errors.OperationError("Unsupported operation")
        elif len(operations) > self.settings.get("max_operations"):
            raise errors.OperationError("Too many operations")

    def _validate_url(self):
        url = self.get_argument("url")
        if not url:
            raise errors.UrlError("Missing url")
        elif url.startswith("http://") or url.startswith("https://"):
            return
        elif self.settings.get("implicit_base_url") and url.startswith("/"):
            return
        raise errors.UrlError("Unsupported protocol")

    def _validate_client(self):
        client = self.settings.get("client_name")
        if client and self.get_argument("client") != client:
            raise errors.ClientError("Invalid client")

    def _validate_signature(self):
        key = self.settings.get("client_key")
        if key and not verify_signature(key, urlparse(self.request.uri).query):
            raise errors.SignatureError("Invalid signature")

    def _validate_host(self):
        hosts = self.settings.get("allowed_hosts", [])
        if hosts and urlparse(self.get_argument("url")).hostname not in hosts:
            raise errors.HostError("Invalid host")


def parse_command_line():  # pragma: no cover
    tornado.options.parse_command_line()


def start_server(app=None):  # pragma: no cover
    if options.debug:
        logger.setLevel(logging.DEBUG)
    server = tornado.httpserver.HTTPServer(
        app if app else PilboxApplication())
    logger.info("Starting server...")
    try:
        server.bind(options.port)
        server.start(1 if options.debug else options.workers)
        tornado.ioloop.IOLoop.instance().start()
    except KeyboardInterrupt:
        tornado.ioloop.IOLoop.instance().stop()


def main(app=None):
    parse_command_line()
    start_server(app)


if __name__ == "__main__":
    main()
